src/coreclr/jit/fgwasm.cpp

```diff
@@ -123,13 +167,805 @@ FlowGraphDfsTree* Compiler::fgWasmDfs()
+        JITDUMP("\n*** %u total Irreducible!\n", numIrreducible);
+    }
+}
+
+//-----------------------------------------------------------------------------
+// WasmFindSccsCore: find strongly connected components in a subgraph
+//
+// Arguments:
+//   subset  - bv describing the subgraph
+//   sccs    - [out] collection of Sccs found in this subgraph
+//   postorder - array of BasicBlock* in postorder
+//   postorderCount - size of the array
+//
+// Notes:
+//   Uses Kosaraju's algorithm: we walk the reverse graph starting
+//   from each block in reverse postorder.
+//
+void FgWasm::WasmFindSccsCore(BitVec& subset, ArrayStack<Scc*>& sccs, BasicBlock** postorder, unsigned postorderCount)
+{
+    SccMap              map(Comp()->getAllocator(CMK_WasmSccTransform));
+    BitVecTraits* const traits = GetTraits();
+
+    // TODO: if we had a BV iter that worked from highest set
+    // bit to lowest, we could iterate the subset directly
+    // and avoid searching here.
+    //
+    for (unsigned i = 0; i < postorderCount; i++)
+    {
+        unsigned const    rpoNum = postorderCount - i - 1;
+        BasicBlock* const block  = postorder[rpoNum];
+
+        if (!BitVecOps::IsMember(traits, subset, block->bbPostorderNum))
+        {
+            continue;
+        }
+
+        AssignBlockToScc(block, block, subset, sccs, map);
+    }
+
+    if (sccs.Height() > 0)
+    {
+        for (int i = 0; i < sccs.Height(); i++)
+        {
+            Scc* const scc = sccs.Bottom(i);
+            scc->Finalize();
+        }
+    }
+}
+
+//-----------------------------------------------------------------------------
+// AssignBlockToScc: assign block to an SCC, then recursively assign its predecessors
+//
+// Arguments:
+//   block   - block to assign
+//   root    - root block of the SCC
+//   subset  - bv describing the subgraph
+//   sccs    - [out] collection of Sccs found in this subgraph
+//   map     - map from block to SCC
+//
+// Notes:
+//   Most blocks won't be in an Scc. So initially we map a block to a null entry in the map.
+//   If we then get a second block in that Scc, we allocate an Scc instance and add both blocks.
+//
+//   This does a reverse-graph walk. Consider expressing this non-recursively.
+//
+void FgWasm::AssignBlockToScc(BasicBlock* block, BasicBlock* root, BitVec& subset, ArrayStack<Scc*>& sccs, SccMap& map)
+{
+    BitVecTraits* const traits = GetTraits();
+
+    // Ignore blocks not in the subset
+    //
+    // This might be too restrictive. Consider
+    //
+    // X -> A; X -> B;
+    // A -> B; A -> C;
+    // B -> A; B -> B;
+    //
+    // We find the A,B Scc. Its non-header subset is empty.
+    //
+    // Thus we fail to find the B self loop "nested" inside.
+    //
+    // Might need to be: "ignore in-edges from outside the set, or from
+    // non-dominated edges in the set...?" So we'd ignore A->B and B->A,
+    // but not B->B.
+    //
+    // However I think we still find all nested Sccs, since those cannot
```

Comment:
Do you think this needs any further investigation to ensure correctness? Is it worth opening a follow-up issue or tracking this somewhere to ensure the case is properly handled?